core/src/main/java/org/apache/iceberg/TableMetadata.java

```diff
@@ -481,9 +488,31 @@ public Snapshot currentSnapshot() {
   }
 
   public List<Snapshot> snapshots() {
+    ensureSnapshotsLoaded();
+
     return snapshots;
   }
 
+  private void ensureSnapshotsLoaded() {
```

Comment:
This should probably be `synchronized` because it is okay to share tables across threads.